File: Simple_Linear_Regression.py
"""
Created on Mon Sep 25 00:32:40 2017
@author: mahmoud
"""


import numpy as np  #mathematical tools"""
import matplotlib.pyplot as plt  #sub library of matplotlib, help us plot nice charts"""
import pandas as pd 

#-- Importing the dataset

dataset = pd.read_csv('Salary_Data.csv')    #2 columns YearsExperience and Salary

                                            # Years Experienc
X = dataset.iloc[:, :-1].values             # [ Lines , Columns ] of data set ':' means all , ':-1' means all columns except last one

                                            # Salary
y = dataset.iloc[:, 1].values               # 1 is the number of the column with Years Experienc being no. 0


#-- Splitting the dataset into the Training set and Test set
from sklearn.cross_validation import train_test_split

                                             #test size is 20% of X and y
                                             #random_state we will have random results if we dont put it
                                             #we have 30 example we will make 20 as training and 10 as testing set 

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 1/3, random_state = 0)

"""                    
#-- Feature Scaling
                    #we are scaling the age and salary becase they are not in the same level
                    #and sometimes we will need to do it for decreasing the run time like in decission tress it will run for a long time if we didn't do it
                    #we don't need to apply feature scaling to y for now bec. they are only 0 and 1

from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()                  #sc for scale, object of the class
X_train = sc_X.fit_transform(X_train)    #when applying scalling to your training set you have to fit the scalling set and then transform it
X_test = sc_X.transform(X_test)          #dont need to fit it bec. its alreay fit in train set
                        #All the Varibales will be from -1 to +1
"""

# Fitting simple Linear Regression to the training set, make the machine learn the relation between yearExperience and salary
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()              #create object of the class
regressor.fit(X_train, y_train)             #fit the regressor item into the training set


# Predicting the Test set results
y_pred = regressor.predict(X_test)           #the vector of predictions of the dependant variable


# Visualising the training set results
plt.scatter(X_train, y_train, color = 'red')    #make points in graph, x coordinate then y then choose color
plt.plot(X_train, regressor.predict(X_train), color = 'blue')      #make line, y coordi = see how he could have predicted our red points, predict how he predict our already made red points
plt.title('Salary vs Experience (Training set)')    #just improve its UI
plt.xlabel('Years of Experience')
plt.ylabel('Salary')
plt.show()                                      #end of graph and ready to plot it



# Visualising the test set results
plt.scatter(X_test, y_test, color = 'red')    #make points in graph, x coordinate then y then choose color
plt.plot(X_train, regressor.predict(X_train), color = 'blue')     #we should not change it with test because we made it by the train set(20 example) #make line, y coordi = see if we predicted our test set as how it should have been?
plt.title('Salary vs Experience (Test set)')    #just improve its UI
plt.xlabel('Years of Experience')
plt.ylabel('Salary')
plt.show()                                      #end of graph and ready to plot it











#y coordi = will be the something we are trying to guess

File: Simple_Linear_Regression.R
# Data Preprocessing

#-- Importing the dataset
dataset = read.csv('Salary_Data.csv')
#dataset = dataset[, 2:3] subset of it


#-- Taking care of missing data
#-- Encoding categorical data
                    #will use factor function 

#-- Splitting the dataset into the Training set and Test set

# install.packages('caTools') comment it just download it once
library(caTools)  #use the library by code or check it from packages tab 
set.seed(123)     #use seed to get same result, like we used random_state in python

split = sample.split( dataset$Salary, SplitRatio = 2/3) #put only y in python we had to put X and y
                                                  #we hve to be careful in python we put the % of test set but here
                                                  #we have to put it for the train set
                                                  #will return true or false for each observation so each one will have true or false
                                                  #True observation goes to train set and False goes to test set

training_set = subset(dataset, split == TRUE)  #subset of Our dataset, who's values are TRUE
test_set = subset(dataset, split == FALSE)     #subset of Our dataset who's values are FALSE



#-- Feature Scaling                       #scale is already programmed function
{
training_set[, 2:3] = scale(training_set[,2:3])        #country and purschased are not numiric by default so we hve to specify the age and salary to scale them
test_set[, 2:3] = scale(test_set[, 2:3])
}

#Fitting Simple Linear Regression to the training set
regressor = lm(formula = Salary ~ YearsExperience,     #1st argum: this means Salary depends on YearsExperience proportinal 
               data = training_set)                    #2nd argum: the data we will train the regressor on


#Predicting the test set results
y_pred = predict(regressor, newdata = test_set) #1st argu: put our regressor, 2nd argu: newdata and that new data is what we want to predit/test its result

